packages/cache/src/CacheStorage.ts

```diff
@@ -1,56 +1,75 @@
 import { ICacheStorage } from "backfill-config";
 
-const savedHashOfRepos: { [gitRoot: string]: { [file: string]: string } } = {};
+const savedHashes: Map<string, Map<string, string>> = new Map();
+
+// Make this feature opt-in as it has not get been tested at scale
+const excludeUnchanged = process.env["BACKFILL_EXCLUDE_UNCHANGED"] === "1";
+
+// Input and output path are absolute (with platform specific separators)
+// For simplicity's sake we assume that we don't have nested git repos
+const foundRoots: Set<string> = new Set();
+async function getRepoRoot(cwd: string): Promise<string> {
+  for (const root of foundRoots.values()) {
+    if (cwd.startsWith(root)) {
+      return root;
+    }
+  }
 
-function getRepoRoot(cwd: string): string {
   // .git is typically a folder but will be a file in a worktree
   const nearestGitInfo =
-    findUp.sync(".git", { cwd, type: "directory" }) ||
-    findUp.sync(".git", { cwd, type: "file" });
+    (await findUp(".git", { cwd, type: "directory" })) ||
+    (await findUp(".git", { cwd, type: "file" }));
   if (!nearestGitInfo) {
     throw new Error(
       "The location that backfill is being run against is not in a git repo"
     );
   }
 
-  return dirname(nearestGitInfo);
+  const result = dirname(nearestGitInfo);
+  foundRoots.add(result);
+  return result;
 }
 
-function fetchHashesFor(cwd: string) {
-  const gitRoot = getRepoRoot(cwd);
-
-  savedHashOfRepos[gitRoot] ||
-    (savedHashOfRepos[gitRoot] = getPackageDeps(gitRoot).files);
-}
-
-function getMemoizedHashesFor(cwd: string): { [file: string]: string } {
-  fetchHashesFor(cwd);
-
-  const gitRoot = getRepoRoot(cwd);
-
-  const savedHashOfThisRepo = savedHashOfRepos[gitRoot] as {
-    [file: string]: string;
-  };
-
-  const pathRelativeToRepo = relative(gitRoot, cwd);
-
-  const filesInCwd = Object.keys(savedHashOfThisRepo).filter(
-    (o) => !relative(pathRelativeToRepo, o).startsWith("..")
+// contract: cwd should be absolute
+// The return keys are relative path with posix file separators
+async function getGitHashesFor(cwd: string): Promise<Map<string, string>> {
+  const result = new Map<string, string>();
+  // gitRoot is an absolute path (with platform specific separators)
+  const gitRoot = await getRepoRoot(cwd);
+  const files = getPackageDeps(gitRoot).files;
+  Object.keys(files).forEach((f) => {
+    // f is a relative path with posix separator
+    const abs = path.join(gitRoot, f);
+    if (abs.startsWith(cwd)) {
+      result.set(path.relative(cwd, abs).split(sep).join("/"), files[f]);
+    }
+  });
+
+  const allFiles = await globby(["**/*", "!node_modules"], { cwd });
+  //globby returns relative path with posix file separator
+  await Promise.all(
+    allFiles.map(async (f) => {
+      if (result.has(f)) {
+        return;
+      }
+      const fileBuffer = await fs.readFile(path.join(cwd, f));
+      const hashSum = crypto.createHash("sha256");
+      hashSum.update(fileBuffer);
```

Comment:
Do we need to think about platform specific file endings?

Reply:
If the build changes only the line ending to a file, we want to pick these up as file change.